source/auth/auth.rst

```diff
@@ -120,6 +120,33 @@ Drivers MUST consider a server ``Unknown`` if authentication fails. Effectively,
 	#. A driver MUST perform authentication with all supplied credentials for all server types with the exception of RSArbiter.
 	#. A single invalid credential is the same as all credentials being invalid.
 
+Mechanism Negotiation via Handshake
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:since: 4.0
+
+If a user provides a username but does not provide an authentication mechanism,
+drivers MUST issue an ``isMaster`` command requesting a user's supported SASL
+mechanisms::
+
+    {isMaster: 1, saslSupportedMechs: "<dbname>.<username>"}
+
+In this example ``<dbname>`` is the authentication database name (default
+'admin') and ``<username>`` is the username provided in the auth credential.
+The username MUST NOT be modified from the form provided by the user (i.e.  do
+not normalize with SASLprep), as the server uses the raw form to look for
+conflicts with legacy credentials.
+
+If the ``isMaster`` command fails, drivers must consider authentication to have
+failed.  If the ``isMaster`` command succeeds, then if the response includes a
+``saslSupportedMechs`` field, then drivers MUST use the contents of that field
+to select a default mechanism as described later.  If that field does not
+exist, drivers MUST use the legacy default mechanism rules for servers older
+than 4.0.
+
+Drivers SHOULD discover a user's supported mechanisms as part of the initial
+handshake on a new connection to avoid an extra database round trip when
+authenticating.
```

Comment:
I suggest that we tighten this up to promote consistency among drivers. When a driver calls isMaster solely for server discovery and monitoring, it MUST NOT include the saslSupportedMechs parameter. When a driver calls isMaster on a connection that will be used for application operations, and the user has provided a username but no authentication mechanism, the driver MUST include the saslSupportedMechs parameter.

Reply:
If we tighten the language I think we should mention that it is not required for drivers that allow multiple users to be authenticated. Eg, if there are two users which one should we send in the isMaster handshake? Such drivers (like pymongo) should do the simple thing and just send an extra isMaster call.

Reply:
I think I'll make two sections.  One for single-auth (MUST do in handshake) and one for multi-auth (MUST do extra roundtrip for each user if no mech specified).  How does that sound?

Reply:
Sounds good. Or if you don't feel like specifying the latter, you could say that if you continue supporting the deprecated multi-auth feature in the SCRAM-SHA-256 era then you're on your own. I'd concur with either choice.

Reply:
Done (added sections).